### PR TITLE
Change the install script and `update` command to use `git ls-remote`

### PR DESCRIPTION
**Pull request by ezraberch**
Based on the discussion in #3958

### Short description 📝

This PR changes how tuist determines the latest version. Rather than parsing the `Constants.swift` (for the install script) or the Changelog (for `update`), the `git ls-remote` command is now used.

### How to test the changes locally 🧐

Run the install script and the `update` command and verify that the latest version is being detected.

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [x] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [x] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [ ] In case the PR introduces changes that affect users, the documentation has been updated.
- [ ] In case the PR introduces changes that affect how the cache artifact is generated starting from the `TuistGraph.Target`, the `Constants.cacheVersion` has been updated.

### CODE REVIEW

File: Sources/TuistEnvKit/Settings/VersionProvider.swift [hunk 1]
```diff
@@ -16,98 +16,63 @@ protocol VersionProviding {
...
         switch self {
-        case let .dataDecodingError(url):
-            return "Error decoding the response from \(url.absoluteString) as an utf8 string."
-        case let .responseError(url, content, statusCode):
-            return """
-            The request to \(url.absoluteString) return an unsuccessful response with status code \(statusCode) and error body:
-            \(content)
-            """
+        case .noVersionsError:
+            return "Error fetching versions from GitHub."
```
Comment (danyf90) [inline Sources/TuistEnvKit/Settings/VersionProvider.swift:24]: The code has no knowledge that it's actually GitHub so I would use git
```suggestion
            return "Error fetching versions from git."
```

File: Sources/TuistSupport/Utils/GitHandler.swift [hunk 2]
```diff
@@ -63,11 +71,23 @@ public final class GitHandler: GitHandling {
         }
     }
 
+    public func lsremote(url: String, tagsOnly: Bool = false, sort: String = "") throws -> String {
+        try capture(command: "git", "ls-remote", tagsOnly ? "-t" : "", !sort.isEmpty ? "--sort=\(sort)" : "", url)
+    }
```
Comment (danyf90) [inline Sources/TuistSupport/Utils/GitHandler.swift:76]: Do we actually need this to be so configurable? I would probably only keep the url and hardcode the others until we really need them. For example the `VersionProvider.swift` should not need to know about host to sort the stuff with the git command (i.e. the "v:refname")

File: Tests/TuistEnvKitTests/Settings/SettingsTests.swift [hunk 3]
```diff
@@ -1,5 +1,57 @@
...
+    }
+
+    override func tearDown() {
+        gitHandler = nil
+        subject = nil
+
+        super.tearDown()
+    }
+
+    func test_highest_remote_version() throws {
```
Comment (danyf90) [inline Tests/TuistEnvKitTests/Settings/SettingsTests.swift:24]:
```suggestion
    func test_latest_remote_version() throws {
```

**Review by danyf90 — CHANGES_REQUESTED** [review]
Thanks for fixing this 👏 👏 👏

**Review by danyf90 — APPROVED** [review]

**Comment by danyf90** [comment]
Even if it's not a user facing change, it would be worthy to add a changeling entry for it

File: Sources/TuistSupport/Utils/GitHandler.swift [hunk 4]
```diff
@@ -63,11 +73,23 @@ public final class GitHandler: GitHandling {
...
+
     private func run(command: String...) throws {
         if Environment.shared.isVerbose {
             try system.runAndPrint(command, verbose: true, environment: System.shared.env)
         } else {
             try system.run(command)
         }
     }
+
+    private func capture(command: String...) throws -> String {
```
Comment (fortmarek) [inline Sources/TuistSupport/Utils/GitHandler.swift:88]: why is this necessary? I think this should be done in `System.swift`, no?

File: Sources/TuistEnvKit/Settings/VersionProvider.swift [hunk 5]
```diff
@@ -16,98 +16,63 @@ protocol VersionProviding {
...
 
-    func changelogResource() -> HTTPResource<String, Error> {
-        resource(path: "/CHANGELOG.md")
-    }
-
     // MARK: - Fileprivate
 
-    fileprivate func parseVersionsFromChangelog(_ changelog: String) throws -> [Version] {
-        let regex = try NSRegularExpression(pattern: "##\\s+([0-9]+.[0-9]+.[0-9]+)", options: [])
+    fileprivate func parseVersionsFromGit(_ gitOutput: String) throws -> [Version] {
```
Comment (fortmarek) [inline Sources/TuistEnvKit/Settings/VersionProvider.swift:64]: What do you think about doing the parsing in `GitHandler` itself? Instead of `String`, it could already return a parsed data since there's not a lot of value of just capturing the raw output.

File: Tests/TuistEnvKitTests/Settings/SettingsTests.swift [hunk 6]
```diff
@@ -1,5 +1,57 @@
...
+        XCTAssertEqual(versions, ["1.9.0", "2.0.0", "2.0.1"])
+    }
+
+    func test_error() throws {
+        gitHandler.lsremoteStub = ""
+        do {
+            _ = try subject.latestVersion().toBlocking().first
+            XCTFail()
+        } catch let error as VersionProviderError {
+            XCTAssertEqual(error.description, "Error fetching versions from git.")
```
Comment (fortmarek) [inline Tests/TuistEnvKitTests/Settings/SettingsTests.swift:52]: I'd recommend using `XCTAssertThrowsSpecific` instead to check for the error.

File: Sources/TuistEnvKit/Settings/VersionProvider.swift [hunk 7]
```diff
@@ -16,98 +16,63 @@ protocol VersionProviding {
...
-                    return AnyPublisher(value: versions)
-                } catch {
-                    return AnyPublisher(error: error)
-                }
-            }
-            .eraseToAnyPublisher()
+        do {
+            let content = try gitHandler.lsRemote(url: Constants.gitRepositoryURL)
+            let versions = try parseVersionsFromGit(content)
+            return AnyPublisher(value: versions)
```
Comment (fortmarek) [inline Sources/TuistEnvKit/Settings/VersionProvider.swift:46]: I don't think there is a point of using `AnyPublisher` here? It should either be converted to actually do the work asynchronously or we can just keep synchronous and remove `AnyPublisher` completely here.

**Review by fortmarek — APPROVED** [review]
Looking good 👏 just a few small nits 🙂

File: Sources/TuistEnvKit/Settings/VersionProvider.swift (hunk 5 above)
Comment (danyf90) [reply Sources/TuistEnvKit/Settings/VersionProvider.swift:64]: I agree this is more a concern of the GitHandler.
We can probably also expose an higher level `remoteTaggedVersions` which returns the tagged versions, instead of `lsRemote`

File: Tests/TuistEnvKitTests/Settings/SettingsTests.swift (hunk 6 above)
Comment (ezraberch) [reply Tests/TuistEnvKitTests/Settings/SettingsTests.swift]: This is moot with the interface change. Now `latestVersion()` returns an optional, so if no versions are found it will be `nil` rather than throwing an exception.

File: Sources/TuistSupport/Utils/GitHandler.swift (hunk 4 above)
Comment (ezraberch) [reply Sources/TuistSupport/Utils/GitHandler.swift:85]: It's done this way to be consistent with the preceding `run` method. @luispadron, that was your creation, so do you have any thoughts on this?

**Review by danyf90 — APPROVED** [review]

File: script/install [hunk 8]
```diff
@@ -20,11 +20,8 @@ warn() {
   printf "${tty_red}Warning${tty_reset}: %s\n" "$(chomp "$1")"
 }
 
-# The line below extracts the version from the Constants.swift file by parsing the line:
-# public static let version = "1.48.1"
-# Note: We can't hit the API from this script because it has limits that might cause the installation to fail.
-LATEST_VERSION=$(curl --silent "https://raw.githubusercontent.com/tuist/tuist/main/Sources/TuistSupport/Constants.swift" | grep 'version =' | sed -E 's/.*"([^"]+)".*/\1/')
-
+# The line below extracts the version from the list of tags in the Tuist repository.
+LATEST_VERSION=$(git ls-remote -t --sort=v:refname https://github.com/tuist/tuist.git | sed -ne '$s/.*tags\/\(.*\)/\1/p')
```
Comment (hisaac) [inline script/install:24]: I've been wondering, is there a reason we're manually parsing the version number? GitHub allows for downloading the latest version using a static URL. Would it work to just use that in the curl command on line 28?

This URL downloads the latest version of the `tuistenv.zip` from releases: https://github.com/tuist/tuist/releases/latest/download/tuistenv.zip

Comment (ezraberch) [reply script/install:24]: As I understand it, the GitHub latest release is determined by timestamp. So if 3.0.0 were released and then sometime later 2.9.1 were released, that link would point to 2.9.1.

Comment (hisaac) [reply script/install:24]: I created a dummy repo here to test: https://github.com/hisaac/test-latest-release/releases

It looks like as long as the tags follow standard semver format, they get organized by version number, not by creation time. I created the `3.0.0` release/tag first, and the `2.0.0` second, and it correctly put the `2.0.0` release below the `3.0.0` one. And using the "latest" download link, it downloads the `3.0.0` release's binary.

https://github.com/hisaac/test-latest-release/releases/latest/download/release.zip

Comment (danyf90) [reply script/install:24]: Nice! I think this simplifies a lot of stuff 🚀

Comment (ezraberch) [reply script/install:24]: Here's an example of a repo where the latest is not the one with the highest version number: https://github.com/vaadin/vaadin-text-field/releases

Comment (hisaac) [reply script/install:24]: Ah bummer, it does have the "Latest" mapped to a lower number then. I wonder how/why it works that way here, but didn't work that way in my test repo. Anyway, carry on!

**Review by laxmorek — APPROVED** [review]

File: Sources/TuistSupport/Utils/GitHandler.swift (hunk 4 above)
Comment (luispadron) [reply Sources/TuistSupport/Utils/GitHandler.swift:85]: I don't recall why that's needed, it may not actually be required. But not a big deal otherwise.
